### PR TITLE
Bump ruff and add new rules

### PR DESCRIPTION
**Pull request by harshil21**
Adds "TRY", parts of "D", "S", and "DOC".

https://docs.astral.sh/ruff/rules/

### CODE REVIEW

File: docs/auxil/link_code.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #
 #  You should have received a copy of the GNU Lesser Public License
 #  along with this program.  If not, see [http://www.gnu.org/licenses/].
+# ruff: noqa: S607, S603
```
Comment (Bibo-Joshi) [inline docs/auxil/link_code.py:18]: Maybe ignore S for the complete docs directory?

File: telegram/_bot.py [hunk 2]
```diff
@@ -4271,7 +4271,7 @@ async def get_updates(
         self,
         offset: Optional[int] = None,
         limit: Optional[int] = None,
-        timeout: Optional[int] = None,
+        timeout: Optional[int] = None,  # noqa: ASYNC109
```
Comment (Bibo-Joshi) [inline telegram/_bot.py:4274]: What's this about?

File: telegram/ext/_application.py [hunk 3]
```diff
@@ -1227,7 +1227,7 @@ async def __create_task_callback(
                 await self.process_error(update=update, error=exception, coroutine=coroutine)
 
             # Raise exception so that it can be set on the task and retrieved by task.exception()
-            raise exception
+            raise
```
Comment (Bibo-Joshi) [inline telegram/ext/_application.py:1230]: So python automatically raises the latest available exception?

**Review by Bibo-Joshi — COMMENTED** [review]
A few questions:)

File: telegram/ext/_application.py (hunk 3 above)
Comment (harshil21) [reply telegram/ext/_application.py:1230]: yep

File: telegram/_bot.py (hunk 2 above)
Comment (harshil21) [reply telegram/_bot.py:4274]: it tells to use Asyncio.timeout which is wrong here ofc

**Review by Bibo-Joshi — APPROVED** [review]

**Pull request merged**